endpoints/bookstore-grpc/bookstore.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+
+import status
+
+
+class ShelfInfo(object):
+  """The contents of a single shelf."""
+  def __init__(self, shelf):
+    self._shelf = shelf
+    self._last_book_id = 0
+    self._books = Bookstore.BookDict()
+
+class Bookstore(object):
+  """An in-memory backend for storing Bookstore data."""
+
+  class ShelfDict(dict):
```

Comment:
yes, you mentioned that. then you provided 3 options, and I think the option 2 is the best to avoid duplicate catch statement and still returning the useful errors. Do you have a better suggestion to do that without using wrapper dict?